noir_stdlib/src/meta/map.nr

```diff
@@ -0,0 +1,458 @@
+    // docs:start:clear
+    pub fn clear(&mut self) {
+        // docs:end:clear
+        self._table = &[Slot::default()];
+        self._len = 0;
+    }
+
+    // Returns true if the map contains a value for the specified key.
+    // docs:start:contains_key
+    pub fn contains_key(
+        self,
+        key: K
+    ) -> bool
+    where
+        K: Hash + Eq,
+        B: BuildHasher<H>,
+        H: Hasher {
+        // docs:end:contains_key
+        self.get(key).is_some()
+    }
+
+    // Returns true if the map contains no elements.
+    // docs:start:is_empty
+    pub fn is_empty(self) -> bool {
+        // docs:end:is_empty
+        self._len == 0
+    }
+
+    // Returns a BoundedVec of all valid entries in this CtHashMap.
+    // The length of the returned vector will always match the length of this CtHashMap.
+    // docs:start:entries
+    pub fn entries(self) -> [(K, V)] {
+        // docs:end:entries
+        let mut entries = &[];
+
+        for slot in self._table {
+            if slot.is_valid() {
+                // SAFETY: slot.is_valid() should ensure there is a valid key-value pairing here
+                let key_value = slot.key_value().unwrap_unchecked();
+                entries = entries.push_back(key_value);
+            }
+        }
+
+        let msg = f"Amount of valid elements should have been {self._len} times, but got {entries.len()}.";
+        assert(entries.len() == self._len, msg);
+
+        entries
+    }
+
+    // Returns a BoundedVec containing all the keys within this CtHashMap.
+    // The length of the returned vector will always match the length of this CtHashMap.
+    // docs:start:keys
+    pub fn keys(self) -> [K] {
+        // docs:end:keys
+        let mut keys = &[];
+
+        for slot in self._table {
+            if slot.is_valid() {
+                let (key, _) = slot.key_value_unchecked();
+                keys = keys.push_back(key);
+            }
+        }
+
+        let msg = f"Amount of valid elements should have been {self._len} times, but got {keys.len()}.";
+        assert(keys.len() == self._len, msg);
+
+        keys
+    }
+
+    // Returns a BoundedVec containing all the values within this CtHashMap.
+    // The length of the returned vector will always match the length of this CtHashMap.
+    // docs:start:values
+    pub fn values(self) -> [V] {
+        // docs:end:values
+        let mut values = &[];
+
+        for slot in self._table {
+            if slot.is_valid() {
+                let (_, value) = slot.key_value_unchecked();
+                values = values.push_back(value);
+            }
+        }
+
+        let msg = f"Amount of valid elements should have been {self._len} times, but got {values.len()}.";
+        assert(values.len() == self._len, msg);
+
+        values
+    }
+
+    // For each key-value entry applies mutator function.
+    // docs:start:iter_mut
+    unconstrained pub fn iter_mut(
+        &mut self,
+        f: fn(K, V) -> (K, V)
```

Comment:
Are we able to use a more Rust-y type here, viz. `f: fn(&mut K, &mut V)`?

Reply:
I don't think so. Since you can't retrieve a mutable reference to an element in an array in Noir.